### PR TITLE
43 charlson erdiag

### PR DESCRIPTION
**Pull request by loffleraSMH**
Closes issue #43 

Allows users to exclude ER diagnoses by setting `erdiag` to `NULL`. Clarified this in the documentation and added a warning message when users don't include ER diagnoses.

Also added input check for `ipdiag` and `erdiag` inputs using `check_input` function.

Note: There may be a merge conflict due to other changes to this function related to issue #48.

### CODE REVIEW

**Comment by vaakesan-SMH** [comment]
I think that this might need a refactor after the changes to the comorbidity functions made in the Elixhauser PR (#48).

**Comment by loffleraSMH** [comment]
> I think that this might need a refactor after the changes to the comorbidity functions made in the Elixhauser PR (#48).

Thanks @vaakesan-SMH! I've made some updates based on the changes in the Elixhauser PR. It's ready for review now.

File: R/comorbidity_index.R [hunk 1]
```diff
@@ -154,13 +164,36 @@ diagnoses_at_admission <- function(ipdiag, erdiag) {
...
+  ### check that ipdiag/erdiag contains genc_id & diagnosis_code
+  check_input(ipdiag, c("data.table", "data.frame"),
+              colnames = c("genc_id", "diagnosis_code"),
+              coltypes = c("", "character"))
+
+  if (!is.null(erdiag)){
+    check_input(erdiag, c("data.table", "data.frame"),
+                colnames = c("genc_id", "er_diagnosis_code"),
+                coltypes = c("", "character"))
+  }
```
Comment (vaakesan-SMH) [inline R/comorbidity_index.R:182]: this is great!

File: R/comorbidity_index.R [hunk 2]
```diff
@@ -6,13 +6,17 @@
 #'
 #' Specifically page 19 for inclusion/exclusion criteria.
 #'
-#' @param ipdiag (`data.table`)\cr
+#' @param ipdiag (`data.table` or `data.frame`)\cr
```
Comment (vaakesan-SMH) [inline R/comorbidity_index.R:9]: Nice catch

File: R/comorbidity_index.R [hunk 3]
```diff
@@ -116,7 +122,11 @@ diagnoses_at_admission <- function(ipdiag, erdiag) {
     .[, exclude := NULL]
 
   ### Combine er-diagnoses and inpatient-diagnoses as diagnoses at admission
-  res <- rbind(ipcharl[, -c("genc_diag")], erdiag[, c("genc_id", "diagnosis_code", "diagnosis_type")])
+  if (!is.null(erdiag)){
+    res <- rbind(ipcharl[, -c("genc_diag")], erdiag[, c("genc_id", "diagnosis_code", "diagnosis_type")])
+  } else {
+    res <- ipcharl[, -c("genc_diag")]
+  }
```
Comment (vaakesan-SMH) [inline R/comorbidity_index.R:129]: This is very much a nitpick that might just come down to personal preference. In this block of logic, we always have to process `ipcharl`, but it occurs in the `else` block, so when reading, it looks like the processing of `erdiag` gets precedence? Also it is a little cleaner to do it this way:

```suggestion
  res <- ipcharl[, -c("genc_diag")]
  if (!is.null(erdiag)){
    res <- rbind(res, erdiag[, c("genc_id", "diagnosis_code", "diagnosis_type")])
  }
```

File: R/comorbidity_index.R [hunk 4]
```diff
@@ -154,13 +164,36 @@ diagnoses_at_admission <- function(ipdiag, erdiag) {
...
+  if (!is.null(erdiag)){
+    erdiag <- coerce_to_datatable(erdiag)
+
+    all_diagnoses <- rbind(
+      erdiag[, .(genc_id, "diagnosis_code" = er_diagnosis_code, "diagnosis_type" = er_diagnosis_type)],
+      ipdiag[, .(genc_id, diagnosis_code, diagnosis_type)]
+    )
+  } else {
+    all_diagnoses <- ipdiag[, .(genc_id, diagnosis_code, diagnosis_type)]
+  }
```
Comment (vaakesan-SMH) [inline R/comorbidity_index.R:196]: I have a similar style comment here as above. But please feel free to take or leave based on your judgement. It has no real consequence.

**Review by vaakesan-SMH — APPROVED** [review]
Looks great. Small nit but feel free to take or leave.

I might also suggest somehow demonstrating this new behaviour in the function tests. Usually in the `diagnoses_at_admission` tests, we supply an empty `erdiag` dataframe (the tests are mostly focused on the `ipdiag` rules). Perhaps in one of the tests, we can change this empty dataframe to a `NULL`, like this:

```
test_that("exclude type 6 and include MRDx if not also type 2", {
  ipdiag <- data.table(
    genc_id = rep(11111111, times = 2),
    diagnosis_code = c("K766", "D529"),
    diagnosis_type = c("M", 6)
  )

  ercols <- c("genc_id", "er_diagnosis_code", "er_diagnosis_type")
  erdiag <- NULL ## <----- HERE (used to be data.table(matrix(nrow = 0, ncol = length(ercols)))
  colnames(erdiag) <- ercols

  res <- diagnoses_at_admission(ipdiag, erdiag)

  expect_equal(
    data.table(
      genc_id = c(11111111),
      diagnosis_code = c("K766"),
      diagnosis_type = c("M")
    ),
    res
  )

  ipdiag <- data.table(
    genc_id = rep(11111111, times = 3),
    diagnosis_code = c("K766", "D529", "K766"),
    diagnosis_type = c("M", 6, 2)
  )

  res <- diagnoses_at_admission(ipdiag, erdiag)

  expect_equal(nrow(res), 0)
})
```

File: R/comorbidity_index.R (hunk 4 above)
Comment (loffleraSMH) [reply R/comorbidity_index.R]: Great, point! I agree this is much nicer to read. I've accepted the change you suggested above and also made a similar change here.

**Comment by loffleraSMH** [comment]
> Looks great. Small nit but feel free to take or leave.
> 
> I might also suggest somehow demonstrating this new behaviour in the function tests. Usually in the `diagnoses_at_admission` tests, we supply an empty `erdiag` dataframe (the tests are mostly focused on the `ipdiag` rules). Perhaps in one of the tests, we can change this empty dataframe to a `NULL`, like this:
> 
> ```
> test_that("exclude type 6 and include MRDx if not also type 2", {
>   ipdiag <- data.table(
>     genc_id = rep(11111111, times = 2),
>     diagnosis_code = c("K766", "D529"),
>     diagnosis_type = c("M", 6)
>   )
> 
>   ercols <- c("genc_id", "er_diagnosis_code", "er_diagnosis_type")
>   erdiag <- NULL ## <----- HERE (used to be data.table(matrix(nrow = 0, ncol = length(ercols)))
>   colnames(erdiag) <- ercols
> 
>   res <- diagnoses_at_admission(ipdiag, erdiag)
> 
>   expect_equal(
>     data.table(
>       genc_id = c(11111111),
>       diagnosis_code = c("K766"),
>       diagnosis_type = c("M")
>     ),
>     res
>   )
> 
>   ipdiag <- data.table(
>     genc_id = rep(11111111, times = 3),
>     diagnosis_code = c("K766", "D529", "K766"),
>     diagnosis_type = c("M", 6, 2)
>   )
> 
>   res <- diagnoses_at_admission(ipdiag, erdiag)
> 
>   expect_equal(nrow(res), 0)
> })
> ```

Thanks @vaakesan-SMH for the review and suggestions! I've made the changes you suggested and also updated one of the tests to include a scenario with `erdiag = NULL`.

**Comment by vaakesan-SMH** [comment]
> Thanks @vaakesan-SMH for the review and suggestions! I've made the changes you suggested and also updated one of the tests to include a scenario with `erdiag = NULL`.

Looks great, feel free to go ahead with the merge
